File: BackTracking/Permutations.py
# Problem Link :- https://leetcode.com/problems/permutations/
'''Given a collection of distinct integers, return all possible permutations.'''

class Solution(object):
    def permute(self, nums):
        n = len(nums)
        if n == 1:
            return [nums]
        res = []
        def helper(arr,path):
            if len(arr) == 2:
                res.append(path+arr)
                res.append(path+arr[::-1])
            for i in range(len(arr)):
                helper([num for num in arr if num != arr[i]], path + [arr[i]])
        helper(nums,[])
        return res


        

File: BackTracking/Permutations.cpp
//Problem Link:- https://leetcode.com/problems/permutations/

#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<int>> ans;
    
    void permute(vector<int>& nums, int index){
        if (index >= nums.size()){
            ans.push_back(nums);
            return;
        }
        for(int i=index;i<nums.size();i++){
            swap(nums[index],nums[i]);
            permute(nums,index+1);
            swap(nums[i],nums[index]);
        }
    }
        
    vector<vector<int>> permute(vector<int>& nums) {
        permute(nums,0);
        return ans;
    }
};